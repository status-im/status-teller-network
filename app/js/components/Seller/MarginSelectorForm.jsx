import React, {Component} from 'react';
import PropTypes from 'prop-types';
import {FormGroup, Label, InputGroup, InputGroupAddon, ButtonGroup, Button, InputGroupText} from 'reactstrap';
import Input from 'react-validation/build/input';
import {withNamespaces} from 'react-i18next';
import Form from 'react-validation/build/form';
import {isNumber, required} from '../../validators';
import MarketButton from '../ui/MarketButton';

const ABOVE = 1;
const BELOW = -1;

class MarginSelectorForm extends Component {
  changeAboveOrBelow = (value) => {
    this.props.onMarginChange({isAbove: value === ABOVE});
  };

  onInputChange = (e) => {
    this.props.onMarginChange({rate: e.target.value});
  };

  render() {
    const {t, fiat, margin} = this.props;

    return (
      <Form ref={c => { this.form = c; }}>
        <h2>{t('sellerMarginContainer.title')}</h2>
        <FormGroup>
            <Label for="margin">{t('marginSelectorForm.margin')}</Label>
            <InputGroup className="full-width-input">
              <Input type="number" name="margin" id="margin" placeholder="0" className="form-control prepend"
                  value={margin.rate} onChange={this.onInputChange} validations={[required, isNumber]} />
              <InputGroupAddon addonType="append"><InputGroupText>%</InputGroupText></InputGroupAddon>
            </InputGroup>
        </FormGroup>
        <FormGroup>
          <ButtonGroup className="d-flex">
            <MarketButton onClick={() => this.changeAboveOrBelow(ABOVE)} active={margin.isAbove}>
              {t('marginSelectorForm.aboveMarket')}
            </MarketButton>
            <MarketButton onClick={() => this.changeAboveOrBelow(BELOW)} active={!margin.isAbove}>
              {t('marginSelectorForm.belowMarket')}
            </MarketButton>
          </ButtonGroup>
        </FormGroup>

        <h3>{t('marginSelectorForm.sellPrice')}</h3>
        <div>
          1 TODO = 1,234.00 {fiat.id}
        </div>
        <small>{t('marginSelectorForm.priceOrigin')}</small>

        <h3>{t('marginSelectorForm.ourFee')}</h3>
        <div>
          TODO SNT
        </div>

        {!this.props.margin.rate && this.props.margin.rate !== 0 && <p className="text-info">{t('marginSelectorForm.enterMargin')}</p>}
      </Form>
    );
  }
}

MarginSelectorForm.propTypes = {
  t: PropTypes.func,
  margin: PropTypes.object,
  fiat: PropTypes.object,
  onMarginChange: PropTypes.func
};

export default withNamespaces()(MarginSelectorForm);
