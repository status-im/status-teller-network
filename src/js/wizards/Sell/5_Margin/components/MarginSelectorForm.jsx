/* global web3 */
import React, {Component, Fragment} from 'react';
import PropTypes from 'prop-types';
import {FormGroup, InputGroup, InputGroupAddon, ButtonGroup, InputGroupText} from 'reactstrap';
import Input from 'react-validation/build/input';
import {withNamespaces} from 'react-i18next';
import Form from 'react-validation/build/form';
import CheckButton from '../../../../ui/CheckButton';
import {isNumber, required, lowerThan} from '../../../../validators';

const ABOVE = 0;
const BELOW = 1;

class MarginSelectorForm extends Component {

  render() {
    const {t, currency, margin, marketType, token, prices, fee} = this.props;

    const basePrice = prices[token.symbol][currency];
    const marginPrice = (margin || 0) / 100 * basePrice;
    const calcPrice = basePrice + (marketType === ABOVE ? marginPrice : -marginPrice);

    return (
      <Form ref={c => { this.form = c; }}>
        <h2>{t('sellerMarginContainer.title')}</h2>
        <FormGroup className="mb-0">
            <InputGroup className="full-width-input">
              <Input type="number"
                     name="margin"
                     id="margin"
                     placeholder="0"
                     className="form-control prepend"
                     value={margin}
                     onChange={(e) => this.props.marginChange(e.target.value)}
                     validations={[required, isNumber, lowerThan.bind(null, 100)]} />
              <InputGroupAddon addonType="append"><InputGroupText>%</InputGroupText></InputGroupAddon>
            </InputGroup>
        </FormGroup>
        <FormGroup>
          <ButtonGroup vertical className="w-100 marginSelector mb-0">
            <CheckButton active={marketType === ABOVE}
                         onClick={() => this.props.marketTypeChange(ABOVE)}>
              {t('marginSelectorForm.aboveMarket')}
            </CheckButton>
            <CheckButton active={marketType === BELOW}
                         onClick={() => this.props.marketTypeChange(BELOW)}>
              {t('marginSelectorForm.belowMarket')}
            </CheckButton>
          </ButtonGroup>
        </FormGroup>
        <h3>{t('marginSelectorForm.sellPrice')}</h3>
        <div className="border rounded p-3">
          1 {token.symbol} = {calcPrice.toFixed(4)} {currency}
        </div>
        <small>{t('marginSelectorForm.priceOrigin')}</small>

        {(fee || '0') !== '0' && <Fragment>
          <h3 className="mt-4">{t('marginSelectorForm.ourFee')}</h3>
          <div className="border rounded p-3">
            {web3.utils.fromWei(fee, 'ether')} SNT
          </div>
        </Fragment>}

        {!this.props.margin && this.props.margin !== 0 && <p className="text-muted mt-3">{t('marginSelectorForm.enterMargin')}</p>}
      </Form>
    );
  }
}

MarginSelectorForm.propTypes = {
  t: PropTypes.func,
  token: PropTypes.object,
  prices: PropTypes.object,
  fee: PropTypes.string,
  margin: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]),
  marketType: PropTypes.number,
  currency: PropTypes.string,
  marginChange: PropTypes.func,
  marketTypeChange: PropTypes.func
};

export default withNamespaces()(MarginSelectorForm);
