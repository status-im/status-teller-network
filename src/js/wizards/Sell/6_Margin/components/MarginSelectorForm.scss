@import "../../../../../css/variable-overrides";
$slider-height: 10px;

.rc-slider {
  .rc-slider-rail {
    width: 95%;
    height: $slider-height;
    background-color: $secondary;
  }

  .rc-slider-track {
    height: $slider-height;
    background-color: $primary;
  }

  .rc-slider-handle{
    height: 30px;
    width: 30px;
    margin-top: -11px;
    border-color: $primary;
  }
}

.margin-input {
  .input-group-append {
    width: 52px;
  }
}
@media screen and (max-width: 450px) {
  .margin-input {
    .form-control {
      padding: 0.375rem 0.75rem;
    }
  }
}
