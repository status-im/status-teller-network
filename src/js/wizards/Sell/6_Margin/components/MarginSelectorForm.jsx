import React, {Component, Fragment} from 'react';
import PropTypes from 'prop-types';
import {FormGroup, InputGroup, InputGroupAddon, InputGroupText, Col, Row} from 'reactstrap';
import Input from 'react-validation/build/input';
import {withNamespaces} from 'react-i18next';
import Form from 'react-validation/build/form';
import {isNumber, required, lowerEqThan, higherThan} from '../../../../validators';
import Slider from 'rc-slider/lib/Slider';
import 'rc-slider/assets/index.css';
import './MarginSelectorForm.scss';

class MarginSelectorForm extends Component {
  onMarginChange = (value) => {
    this.props.marginChange(value);
  };

  render() {
    const {t, currency, margin, token, prices, feeMilliPercent} = this.props;

    const basePrice = prices[token.symbol][currency];
    const marginPrice = (margin || 0) / 100 * basePrice;
    const calcPrice = basePrice + marginPrice;

    return (
      <Form ref={c => {
        this.form = c;
      }}>
        <h2>{t('sellerMarginContainer.title')}</h2>
        <FormGroup className="mb-0">
          <Row>
            <Col md={9} sm={8} xs={7}>
              <Slider className="mb-3 p-4" min={-99} max={100} defaultValue={0}
                      onChange={(value) => this.onMarginChange(value)} value={margin}/>
            </Col>
            <Col>
              <InputGroup className="full-width-input">
                <Input type="number"
                       name="margin"
                       id="margin"
                       placeholder="0"
                       className="form-control prepend"
                       value={margin}
                       onChange={(e) => this.onMarginChange(e.target.value)}
                       validations={[required, isNumber, lowerEqThan.bind(null, 100), higherThan.bind(null, -100)]}/>
                <InputGroupAddon addonType="append"><InputGroupText>%</InputGroupText></InputGroupAddon>
              </InputGroup>
            </Col>
          </Row>
          <p className="text-muted">A negative margin means you sell assets for less than what they are worth</p>
        </FormGroup>

        <h3>{t('marginSelectorForm.sellPrice')}</h3>
        <div className="border rounded p-3">
          1 {token.symbol} = {calcPrice.toFixed(4)} {currency}
        </div>
        <small>{t('marginSelectorForm.priceOrigin')}</small>

        {(feeMilliPercent || '0') !== '0' && <Fragment>
          <h3 className="mt-4">{t('marginSelectorForm.ourFee')}</h3>
          <div className="border rounded p-3">
            {feeMilliPercent / 1000} %
          </div>
        </Fragment>}

        {!margin && margin !== 0 && <p className="text-muted mt-3">{t('marginSelectorForm.enterMargin')}</p>}
      </Form>
    );
  }
}

MarginSelectorForm.propTypes = {
  t: PropTypes.func,
  token: PropTypes.object,
  prices: PropTypes.object,
  feeMilliPercent: PropTypes.string,
  margin: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number
  ]),
  currency: PropTypes.string,
  marginChange: PropTypes.func
};

export default withNamespaces()(MarginSelectorForm);
