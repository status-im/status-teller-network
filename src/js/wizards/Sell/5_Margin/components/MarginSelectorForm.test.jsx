import React from "react";
import { shallow } from "enzyme";

import MarginSelectorForm from "./MarginSelectorForm";

describe('MarginSelectorForm', () => {
  it('should render correctly', () => {
    const component = shallow(<MarginSelectorForm margin={0}
                                                  token={{symbol: 'SNT'}}
                                                  prices={{'SNT': {'USD': 140}}}
                                                  marketType={0}
                                                  currency={"USD"}
                                                  fee={"0"}
                                                  onMarketTypeChange={() => undefined}
                                                  onMarginChange={() => undefined}/>);
  
    expect(component).toMatchSnapshot();
  });
});
