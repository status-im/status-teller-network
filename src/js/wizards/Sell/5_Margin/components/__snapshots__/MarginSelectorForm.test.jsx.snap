// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarginSelectorForm should render correctly 1`] = `
<Form(t)>
  <h2 />
  <FormGroup
    tag="div"
  >
    <Label
      for="margin"
      tag="label"
      widths={
        Array [
          "xs",
          "sm",
          "md",
          "lg",
          "xl",
        ]
      }
    />
    <InputGroup
      className="full-width-input"
      tag="div"
    >
      <Control(s)
        className="form-control prepend"
        id="margin"
        name="margin"
        onChange={[Function]}
        placeholder="0"
        type="number"
        validations={
          Array [
            [Function],
            [Function],
            [Function],
          ]
        }
        value={0}
      />
      <InputGroupAddon
        addonType="append"
        tag="div"
      >
        <InputGroupText
          tag="span"
        >
          %
        </InputGroupText>
      </InputGroupAddon>
    </InputGroup>
  </FormGroup>
  <FormGroup
    tag="div"
  >
    <ButtonGroup
      className="d-flex"
      role="group"
      tag="div"
    >
      <MarketButton
        active={true}
        onClick={[Function]}
      />
      <MarketButton
        active={false}
        onClick={[Function]}
      />
    </ButtonGroup>
  </FormGroup>
  <h3 />
  <div
    className="border rounded p-3"
  >
    1 
    SNT
     = 
    140.0000
     
    USD
  </div>
  <small />
</Form(t)>
`;
